File: 2025/haskell/3.hs
module Main where

main = do
  inputs <- readFile "../inputs/3.txt"
  putStrLn "Task1:"
  print $ task1 $ parse inputs
  putStrLn "Task2:"
  print $ task2 $ parse inputs

example = "987654321111111\n811111111111119\n234234234234278\n818181911112111"

splitOn :: (Eq a) => a -> [a] -> [[a]]
splitOn _ [] = []
splitOn c xs = first : splitOn c rest
  where
    (first, rest) = splitFirst c xs

splitFirst :: (Eq a) => a -> [a] -> ([a], [a])
splitFirst c xs = (before, drop 1 after)
  where
    (before, after) = span (/= c) xs

parse = splitOn '\n'

count 0 _ = []
count n xs = maxElem : count (n - 1) (tail $ dropWhile (/= maxElem) xs)
  where
    maxElem = maximum $ take (length xs - n + 1) xs

counts n = sum . map (read . count n)

task1 = counts 2

task2 = counts 12

File: 2025/template-hell/src/3.cpp
#include "util.hpp"
#include <cstddef>
#include <print>

using example = string_to_list<"987654321111111\n"
                               "811111111111119\n"
                               "234234234234278\n"
                               "818181911112111">;

template <typename xs>
using parse = split_on<V<'\n'>, xs>;

template <std::size_t n, typename xs>
struct Count {

    using part = take<xs::len - n + 1, xs>;
    static_assert(!std::same_as<part, List<>>);
    using max_elem = maximum<take<xs::len - n + 1, xs>>;
    using Value = prepend<max_elem, typename Count<n - 1, tail<drop_until_eq<max_elem, xs>>>::Value>;
};
template <typename xs>
struct Count<0, xs> {
    using Value = List<>;
};

template <std::size_t n, typename xs>
struct Counts {
    template <typename lst>
    using countn = Count<n, lst>::Value;
    using Value = map<parse_int, map<countn, xs>>;
};
template <std::size_t n, typename xs>
using counts = Counts<n, xs>::Value;

template <typename xs>
using part1 = sum<counts<2, xs>>;
template <typename xs>
using part2 = sum<counts<12, xs>>;

#if 0
constexpr char inputs[] =
{
#embed "../../inputs/3.txt"
};
using inp = string_to_list<inputs>;
#endif


int main()
{
    using t1 = part1<parse<example>>;
    using t2 = part2<parse<example>>;
    std::println("Part1:\n{}\nPart2:{}\n", t1{}, t2{});
}
